### PR TITLE
[primgen] Use verible-verilog-syntax for parsing Verilog code

### PR DESCRIPTION
**Pull request by mglb**
Related issue: https://github.com/lowRISC/opentitan/issues/1720

I've tested the code by comparing `*.sv` and `*.core` files generated by `util/syn_yosys.sh` before and after the change. The only differences are slightly different whitespace/formatting and a lack of a few comments.

### CODE REVIEW

**Comment by imphil** [comment]
Thanks for your work on this, it's nice to see it come together.

Could you keep the existing regex-based parsing for now, and use it as fall-back if a suitable version of Verible isn't found (after warning loudly)? primgen is used by everybody who wants to build OpenTitan hardware, and we need to make sure that it is sufficiently easy for all our users to have Verible installed first and before breaking their builds.

**Comment by mglb** [comment]
@imphil Fallback added.

I'll wait for CI results and switch to ready for review on green.

File: hw/ip/prim/util/primgen.py [hunk 1]
```diff
@@ -196,6 +251,54 @@ def _parse_parameter_port_list(parameter_port_list):
     return parameters
 
 
+def _parse_module_header(generic_impl_filepath, module_name):
```
Comment (imphil) [inline hw/ip/prim/util/primgen.py:254]: We should be able to simplify this function a bit and make it more robust by doing two things:

* Generally, avoid checking for versions, check for features. In this particular case, there might be good reasons to additionally check the version if a call fails, to give users a hint that their Verible version is too old, but do that only for the user if the feature check fails (e.g. try to call Verible with the dumping arguments, and see if it returns an error).

* It feels to me that the compatibility checking code should actually go into the `VeribleVerilogSyntax()` class, throwing an exception if Verible is too old (or other required features are missing). Can you do that?

With those two things together, the code in this function should reduce to something like

```python3
try:
    return _parse_module_header_verible(generic_impl_filepath, module_name)
except Exception as e:
    print(f"Unable to parse file XY with Verible, using regex fallback instead: {e.msg}")
    return _parse_module_header_fallback(generic_impl_filepath, module_name)
```

File: python-requirements.txt [hunk 2]
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # Keep sorted.
+anytree
```
Comment (imphil) [inline python-requirements.txt:6]: Can you add a comment that this dependency is used by the vendored-in  sw/host/vendor/google_verible_verilog_syntax_py code? (To make it easier to remove it once this functionality is available as Python package.)

File: hw/lint/tools/veriblelint/rules.vbl [hunk 3]
```diff
@@ -17,3 +17,5 @@ parameter-name-style=localparam_style:CamelCase|ALL_CAPS
 
 # we allow nested struct definitions
 -typedef-structs-unions
+
+-generate-label-prefix
```
Comment (imphil) [inline hw/lint/tools/veriblelint/rules.vbl:21]: Why is this change needed? Is it part of the Verible version update (if so, let's tackle it in a separate PR or add to #6434) or a requirement for the parsing code?

**Review by imphil — COMMENTED** [review]
Thanks for this pull request @mglb, and sorry for the longer wait. We are now ready to move forward with this, now that the dust of other changes has settled a bit.

I left some comments in line. Additionally:

* We need to do the Verible version update separately. I opened #6434 and will take care of it.
* Please squash the commits touching primgen.py into a single one to give us an "atomic" commit that can be reverted if necessary without leaving the tree in a half-working state.

File: hw/lint/tools/veriblelint/rules.vbl (hunk 3 above)
Comment (mglb) [reply hw/lint/tools/veriblelint/rules.vbl]: When I bumped Verible version the linter in CI has been complaining that some Opentitan code violates this (newly introduced) rule. This seems to be fixed now.

File: python-requirements.txt (hunk 2 above)
Comment (mglb) [reply python-requirements.txt]: Done

File: hw/ip/prim/util/primgen.py (hunk 1 above)
Comment (mglb) [reply hw/ip/prim/util/primgen.py:271]: Partially done. The only missing thing is reporting that the Verible version is too old - I'm moving that part to upstream's VeribleVerilogSyntax class. The change prototyped on Opentitan copies of the class: https://github.com/antmicro/opentitan/commit/a0c6489aa1717878431aaf3b87bf7a85ce81214d

File: hw/ip/prim/util/primgen.py [hunk 4]
```diff
@@ -196,6 +251,54 @@ def _parse_parameter_port_list(parameter_port_list):
...
+        actual_semver = verible_req.to_semver(actual_sv, False)
+
+        if min_semver <= actual_semver:
+            return _parse_module_header_verible(generic_impl_filepath,
+                                                module_name)
+
+        warning_reason = 'Verible version ({}) is older than required ({})'.format(
+            actual_sv, MIN_VERIBLE_VERSION)
+    except Exception:
+        warning_reason = 'Failed to get Verible version.'
```
[the remark this answers is not in the document]
Comment (mglb) [reply hw/ip/prim/util/primgen.py]: Done.

File: hw/ip/prim/util/primgen.py [hunk 5]
```diff
@@ -10,6 +10,17 @@
...
 from mako.template import Template
 
+_SELF_DIR = os.path.dirname(__file__)
+_OPENTITAN_DIR = os.path.normpath(os.path.join(_SELF_DIR, '../../../../'))
+_UTIL_DIR = os.path.join(_OPENTITAN_DIR, 'util/')
+_VERIBLE_VERILOG_SYNTAX_PY_DIR = os.path.join(
+    _OPENTITAN_DIR, 'sw/host/vendor/google_verible_verilog_syntax_py/')
+
+sys.path.append(_UTIL_DIR)
+sys.path.append(_VERIBLE_VERILOG_SYNTAX_PY_DIR)
```
[the remark this answers is not in the document]
Comment (mglb) [reply hw/ip/prim/util/primgen.py:15]: Done

File: hw/ip/prim/util/primgen.py [hunk 6]
```diff
@@ -84,29 +95,72 @@ def _top_module_file(core_files, module_name):
...
-    #   requires Icarus and Python 3.6
-    # - https://kevinpt.github.io/hdlparse/:
-    #   Didn't do anything in my tests.
-    # - https://github.com/sepandhaghighi/verilogparser:
-    #   Doesn't support SystemVerilog
+    from verible_verilog_syntax import VeribleVerilogSyntax, PreOrderTreeIterator
+
+    verible = os.environ.get("VERIBLE_VERILOG_SYNTAX_BINARY",
+                             "verible-verilog-syntax")
+    parser = VeribleVerilogSyntax(executable=verible)
```
[the remark this answers is not in the document]
Comment (mglb) [reply hw/ip/prim/util/primgen.py]: Done

File: hw/ip/prim/util/primgen.py [hunk 7]
```diff
@@ -84,29 +95,72 @@ def _top_module_file(core_files, module_name):
...
+
+    data = parser.parse_file(generic_impl_filepath, options={"skip_null": True})
+    if not data.tree:
+        raise ValueError("Unable to parse %s." %
+                         (generic_impl_filepath, ))
```
[the remark this answers is not in the document]
Comment (mglb) [reply hw/ip/prim/util/primgen.py]: Done

> And it would be much better API if the parse_file() method would throw an parse exception itself. Maybe you can follow-up upstream for that?

The object returned by `parse_file` contains list of errors reported by Verible. They could be returned in an exception object, but:
* There is a small chance that Verible recovered from errors and a tree has been generated.
* There are errors, the tree is not generated, but tokens list is (when requested)


As for list of errors, they are printed now:
```
Inspecting generic module /home/mglb/projects/systemverilog/src/opentitan/hw/ip/prim_generic/rtl/prim_generic_xor2.sv
Verible: parse error in line 12 column 0.
Verible: parse error in line 14 column 2.
Verible: parse error in line 14 column 31.
Unable to parse '/home/mglb/projects/systemverilog/src/opentitan/hw/ip/prim_generic/rtl/prim_generic_xor2.sv'.
(...)
```

File: hw/ip/prim/util/primgen.py [hunk 8]
```diff
@@ -10,6 +10,17 @@
 import yaml
 from mako.template import Template
 
+_SELF_DIR = os.path.dirname(__file__)
+_OPENTITAN_DIR = os.path.normpath(os.path.join(_SELF_DIR, '../../../../'))
```
[the remark this answers is not in the document]
Comment (mglb) [reply hw/ip/prim/util/primgen.py]: Done

**Comment by mglb** [comment]
>  * Please squash the commits touching primgen.py into a single one to give us an "atomic" commit that can be reverted if necessary without leaving the tree in a half-working state.

Done.

> * Can you insert a comment into the generated code how it was generated (regex or verible) to make debugging easier if the process fails?

Done:
```verilog
// This file is auto-generated.
// Used parser: Verible v0.0-1213-g9e5c085
```

**Review by imphil — APPROVED** [review]
LGTM, thanks for your persistent work on this!
